File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextTransform.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.text;

import androidx.annotation.Nullable;
import java.text.BreakIterator;

/** Types of text transforms for CustomTextTransformSpan */
public enum TextTransform {
  NONE,
  UPPERCASE,
  LOWERCASE,
  CAPITALIZE,
  UNSET;

  public static String apply(@Nullable String text, TextTransform textTransform) {
    if (text == null) {
      return null;
    }

    String transformed;
    switch (textTransform) {
      case UPPERCASE:
        transformed = text.toUpperCase();
        break;
      case LOWERCASE:
        transformed = text.toLowerCase();
        break;
      case CAPITALIZE:
        transformed = capitalize(text);
        break;
      default:
        transformed = text;
    }

    return transformed;
  }

  private static String capitalize(String text) {
    BreakIterator wordIterator = BreakIterator.getWordInstance();
    wordIterator.setText(text);

    StringBuilder res = new StringBuilder(text.length());
    int start = wordIterator.first();
    for (int end = wordIterator.next(); end != BreakIterator.DONE; end = wordIterator.next()) {
      res.append(Character.toUpperCase(text.charAt(start)));
      res.append(text.substring(start + 1, end));
      start = end;
    }

    return res.toString();
  }
};

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextTransform.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.text

import java.text.BreakIterator
import java.util.Locale

/** Types of text transforms for CustomTextTransformSpan */
internal enum class TextTransform {

  NONE,
  UPPERCASE,
  LOWERCASE,
  CAPITALIZE,
  UNSET;

  internal companion object {
    @JvmStatic
    fun apply(text: String, textTransform: TextTransform?): String =
        text.applyTextTransform(textTransform)
  }
}

internal fun String.applyTextTransform(textTransform: TextTransform?): String {
  return when (textTransform) {
    TextTransform.UPPERCASE -> uppercase(Locale.getDefault())
    TextTransform.LOWERCASE -> lowercase(Locale.getDefault())
    TextTransform.CAPITALIZE -> {
      val wordIterator = BreakIterator.getWordInstance()
      wordIterator.setText(this)
      val res = StringBuilder(length)
      var start = wordIterator.first()
      var end = wordIterator.next()
      while (end != BreakIterator.DONE) {
        res.append(substring(start, end).replaceFirstChar { it.uppercaseChar() })
        start = end
        end = wordIterator.next()
      }
      res.toString()
    }
    else -> this
  }
}
